### DISCUSSION TITLE
Next js SSR

### DISCUSSION THREAD

**Discussion by Mostafa-Mohammadi** [vercel/next.js#41512] [Help]
We know when we use getServerSideProps on a page, at the request time server creates a HTML and sends it back immediately to the browser,

My confusion is why nevertheless our initial loading is very slow than CSR even without any data fetching on SSR, indeed it takes about 5 seconds to show the page.

Thanks in Advance for any help

**Comment by icyJoseph** [comment]
> initial loading is very slow than CSR even without any data fetching on SSR, indeed it takes about 5 seconds to show the page.

Could you share a link to that page? And well, without data fetching, means no `getServerSideProps`(GSSP) right? then the page should've been pre-rendered at build time, or if you have `_app` with `getInitialProps`, perhaps there's some blocking process there?

Could you describe the Next.js architecture of a page that's taking this long?

**Reply by Mostafa-Mohammadi** [reply]
Actually, I mean I use getServerSideProps and without any data fetching on it, I just return empty props.

But in production, every page takes about 2.5 seconds to show content, is it normal for SSR?

And in your opinion is SSR faster than CSR in any case? if so, could we use SSR for all of the pages?

Because it has less TTFB and doesn't have well performance for navigation between pages.

Thanks.

**Reply by icyJoseph** [reply]
I think the concept of SSR is a bit unclear these days.

- Next.js can generate a page at build time
- Or it can generate it as the request arrives

Both cases hydrate, which means that unlike regular CSR, where the initial frame is an empty shell, in the two cases above, the initial HTML is complete. 

This enabled bots, and blocked-JS entities to see your site, and even navigate between pages, to a certain extend...

Time to visible in Next.js' case is fast when the page is generated at build time, however, Time to interactive is on par with CSR. Because of the hydration process.

A page generated on demand, ignoring fallbacks for ISR, waits for data on the server, then uses the data to render a string and sends that to the client. Time to visible, and interactive are very variable, but in general, the loading time is affected by, size of the JS running the page, size of the data injected to the page and stylesheets. A large page with not much data is still affected by network speeds and parsing, loading and execution of JavaScript on the browser, however, as soon as the HTML arrives, there'll be something to see, in all cases.

For pages with `getServerSideProps`(GSSP) higher latency is to be expected, if many requests arrive at the same time, then there's less memory to serve each, the size of packages used, the size of HTML output, all of this affects your result. 

Note that during SPA transitions, navigating into a GSSP page, gets only the JSON for that page, the JS for it is prefetched by next/link mechanics. So technically it might be perceived as a faster transition than directly navigating into the page.

Last but not least, a CDN could, in some cases, be used to cache GSSP results and entire HTML responses. And there's also ISR which is a Next.js controlled way to nearly do the same.

> And in your opinion is SSR faster than CSR in any case? if so, could we use SSR for all of the pages?

This is a very grey area. I think the next.js docs say it best:

> You should use getServerSideProps only if you need to render a page whose data must be fetched at request time. 

That doesn't mean skip SSR though. Next.js capable of both SSR and CSR, the question with SSR is, when do you do it, at build time? or as the requests arrive.

Ideally you should pre-render as many pages as possible, and keep only a few with GSSP. Some pages might be pre-rendered with empty slots in the HTML and fill those in during client runtime, but even that depends on how the page will be primarily consumed.

**Reply by Mostafa-Mohammadi** [reply]
Thanks for your explanation.
